ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/NullScanTaskDispatcher.java

```diff
@@ -93,94 +91,109 @@ private String getAliasForTableScanOperator(MapWork work,
-        .filter(a -> aliases.contains(a)).collect(Collectors.toList());
-    if (!allowed.isEmpty()) {
-      PartitionDesc partDesc = work.getPathToPartitionInfo().get(path).clone();
-      PartitionDesc newPartition =
-          changePartitionToMetadataOnly(partDesc, path);
-      // Prefix partition with something to avoid it being a hidden file.
-      Path fakePath =
-          new Path(NullScanFileSystem.getBase() + newPartition.getTableName()
-              + "/part" + encode(newPartition.getPartSpec()));
-      StringInternUtils.internUriStringsInPath(fakePath);
-      work.addPathToPartitionInfo(fakePath, newPartition);
-      work.addPathToAlias(fakePath, new ArrayList<>(allowed));
-      aliasesAffected.removeAll(allowed);
-      if (aliasesAffected.isEmpty()) {
-        work.removePathToAlias(path);
-        work.removePathToPartitionInfo(path);
-      }
-    }
+  private void addCandidatePath(Map<Path, Collection<String>> aliasesForPath, Path path, String alias) {
+    Collection<String> aliases = aliasesForPath.computeIfAbsent(path, k -> new ArrayList<>());
+    aliases.add(alias);
   }
 
-  private void processAlias(MapWork work, Set<TableScanOperator> tableScans) {
-    Set<String> aliases = new HashSet<>();
+  private void processTableScans(MapWork work, Set<TableScanOperator> tableScans) {
+    Map<Path, Boolean> managedEmptyPathMap = new HashMap<>();
+    Map<Path, Collection<String>> candidatePathsToAliases = new HashMap<>();
+    Map<String, Boolean> aliasTypeMap = new HashMap<>();
+    Map<String, Map<Path, Partition>> allowedAliasesToPartitions = new HashMap<>();
     for (TableScanOperator tso : tableScans) {
       // use LinkedHashMap<String, Operator<? extends OperatorDesc>>
       // getAliasToWork() should not apply this for non-native table
       if (tso.getConf().getTableMetadata().getStorageHandler() != null) {
         continue;
       }
       String alias = getAliasForTableScanOperator(work, tso);
-      aliases.add(alias);
+      boolean isManagedTable = !MetaStoreUtils.isExternalTable(tso.getConf().getTableMetadata().getTTable());
+      aliasTypeMap.put(alias, isManagedTable);
+      allowedAliasesToPartitions.put(alias, getPathToPartitionMap(alias, tso));
       tso.getConf().setIsMetadataOnly(true);
     }
-    // group path alias according to work
-    Map<Path, List<String>> candidates = new HashMap<>();
+
     for (Path path : work.getPaths()) {
-      List<String> aliasesAffected = work.getPathToAliases().get(path);
-      if (CollectionUtils.isNotEmpty(aliasesAffected)) {
-        candidates.put(path, aliasesAffected);
+      List<String> aliases = work.getPathToAliases().get(path);
+      for (String alias: aliases) {
+        Map<Path, Partition> pathToPartitionMap = allowedAliasesToPartitions.get(alias);
+        if (pathToPartitionMap == null) {
+          continue;
+        }
+        addCandidatePath(candidatePathsToAliases, path, alias);
```

Comment:
Can be candidatePathsToAliases.computeIfAbsent(path, k -> new ArrayList<>()).add(alias) ?

Reply:
makes sense